Trying out adding globalProperties to entire ChildLogger chain

markfields opened the pull request:
Addresses #8069 in a slightly different way than requested and lays groundwork for fixing #7778 (just update container logger to use globalProperties instead of properties)

markfields committed:
Add callsite version to events and errors for logging

Unsure of this approach, trying it out.  Maybe prefix version with namespace from child logger chain?

markfields committed:
Try building up layerVersions prop

markfields committed:
Revert "Add callsite version to events and errors for logging"

This reverts commit 47859a4f738158afde3f4ac76ae6e8a794ebec5f.
Because telemetry-utils pkgVersion may not match
layer package version

markfields committed:
Add test for layerVersions

markfields committed:
Broken! But does something.

markfields committed:
My tests pass but this ugly!

markfields committed:
Yay fixed it

markfields committed:
Cleanup

markfields committed:
let ChildLoggerRoot apply the global props

markfields committed:
Test coverage

markfields committed:
use layerVesrion for loader and odspdriver

markfields committed:
api md

markfields commented on packages/utils/telemetry-utils/src/logger.ts:
```diff
@@ -270,7 +299,14 @@ export class ChildLogger extends TelemetryLogger {
-        properties?: ITelemetryLoggerPropertyBags): TelemetryLogger {
+        properties: ITelemetryLoggerPropertyBags = {},
+        globalProperties: ITelemetryLoggerPropertyBags = {},
+        layerVersion?: string,
```
Thinking about ditching this generic capability and just have loader and driver pass `fooVersion` as a global property.  Thoughts?

anthony-murphy commented on packages/utils/telemetry-utils/src/logger.ts:
```diff
@@ -270,7 +299,14 @@ export class ChildLogger extends TelemetryLogger {
+
         // if we are creating a child of a child, rather than nest, which will increase
         // the callstack overhead, just generate a new logger that includes everything from the previous
         if (baseLogger instanceof ChildLogger) {
```
this doesn't always work across our layers, loader/runtime. it wasn't a problem before as this code is really just an optimization as we ran into perf problems with extereme nesting, but in this case it will be an issue, as global props won't cross this boundary

anthony-murphy replied:
Another idea is to build this directly into the telemetry logger class with a function like `protected addGlobalProps?()` the advantage here is that is would also work with the multisink logger, so we could get the logs into the debug logger. 

Since we don't know if the parent will support globalProps, we should continue to set these props locally. the locall props should override the global props as in the case they both exist it won't matter as the values should match, if they don't, the lowest logger should win

markfields replied:
Ahhh of course.  Thanks, I'll think it over

anthony-murphy commented on packages/utils/telemetry-utils/src/logger.ts:
```diff
@@ -279,42 +315,54 @@ export class ChildLogger extends TelemetryLogger {
+            Object.assign(baseLogger.globalProperties, globalProperties);
+
+            // Concatenate next layer version to existing versions
+            if (namedLayerVersion !== undefined) {
```
i really don't like this. it'll be hard to aggregate. i prefer the version per column approach today

anthony-murphy replied:
at a minimum this should be json, which has support for parsing in most data tools

markfields replied:
Yeah, I'm good to ditch it.

markfields committed:
Struggling with protected

markfields committed:
Merge branch 'main' into log-versions

markfields committed:
Revert "Struggling with protected"

This reverts commit 39c16aeb04bde845a0658cabe20910aef644fa1d.

markfields committed:
Cover case where ChildLogger trees are nested
due to layer boundaries and instanceof semantics.

In this case, ChildLoggerRoot's parent would be a ChildLogger under another root.

vladsud commented on api-report/telemetry-utils.api.md:
```diff
@@ -23,9 +23,13 @@ import { TypedEventEmitter } from '@fluidframework/common-utils';
+    // Warning: (ae-forgotten-export) The symbol "ChildLoggerRoot" needs to be exported by the entry point index.d.ts
+    //
+    // (undocumented)
+    protected readonly rootLogger: ChildLoggerRoot;
```
This seems like limiting logger nesting / dependency to always be straight line (based on glancing at PR - up to a point where it's not / where parent does not satisfy some signature). For example, this would not work for MultiSinkLogger, right?

vladsud commented on api-report/telemetry-utils.api.md:
```diff
@@ -23,9 +23,13 @@ import { TypedEventEmitter } from '@fluidframework/common-utils';
 
 // @public
 export class ChildLogger extends TelemetryLogger {
+    static create(baseLogger?: ITelemetryBaseLogger, namespace?: string, properties?: ITelemetryLoggerPropertyBags, globalProperties?: ITelemetryLoggerPropertyBags, layerVersion?: string): TelemetryLogger;
```
I do not think base layer needs to know anything about concept of layers or versions. I'd prefer these concepts not to be present in base layer (and if needed, to be in add-on layer).

vladsud commented:
Personally, I'd rather not solve this problem (if we can't find better solution), over adding global patterns like this. This makes it much harder to predict / debug/ understand how lower layers impact behavior of higher layers as they can reach out and change values on events emitted by those higher layers I'd rather prefer very explicit things similar to what we are adding to driver - PUSH commination - equivalent of [user agent ](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/userAgent) properties passed in both directions, and explicit place where we add them in telemetry. Driver & runtime can easily provide such things to loader and loader could include such "environment" (another name we use somewhere) in its future events.

markfields committed:
Merge branch 'main' into log-versions

markfields committed:
Support global properties above ChildLoggerRoot

markfields committed:
Remove special layerVersions prop

markfields committed:
api md and test updates

markfields commented:
I totally spaced on supporting `MultiSinkLogger`, of course current approach won't work (need a function like Tony suggested).  Closing this for now, moving #8069 to design-required.

Good feedback, @vladsud, in terms of concern over globalProperties at all.  I think there could be a case to be made still but it's definitely murky.
